### PR TITLE
Law changes

### PR DESCRIPTION
**Pull request by valeriadamante**
this PR is meant to create the skim producer in order to have lighter final anatuples and to make law work.

### CODE REVIEW

File: AnaProd/tasks.py [hunk 1]
```diff
@@ -56,32 +63,27 @@ def create_branch_map(self):
...
-            raise RuntimeError(f'Produced anaTuple {tmp_file} is corrupted')
-        shutil.move(tmp_file, output)
-        print(f'anaTuple for sample {sample_name}  split_idx={split_idx} is created in {output}')
+        outdir_anatuples = os.path.join(job_home, 'anaTuples', sample_name)
+        sh_call([ 'python3', producer_anatuples, '--config', self.sample_config, '--inFile', ','.join(input_files),
+                  '--outDir', outdir_anatuples, '--sample', sample_name, '--anaCache', anaCache, '--customisations',
+                  self.customisations, '--compute_unc_variations', 'True', '--store-noncentral'], env=self.cmssw_env())
+        producer_skimtuples = os.path.join(self.ana_path(), 'Analysis', 'SkimProducer.py')
+        outdir_skimtuples = os.path.join(job_home, 'skim', sample_name)
+        print([ 'python3', producer_skimtuples, '--inputDir',outdir_anatuples, '--workingDir', outdir_skimtuples, '--outputFile', 'skim.root'])
```
Comment (kandrosov) [inline AnaProd/tasks.py:80]:
```suggestion
```
just increase verbosity for sh_call e.g. `verbose=1`

File: AnaProd/tasks.py [hunk 2]
```diff
@@ -116,7 +118,8 @@ def getBranches(samples, central_nanoAOD_path, max_files_per_job):
                     break
         return branches
 
-class AnaTupleTask(Task, HTCondorWorkflow, law.LocalWorkflow):
+'''
```
Comment (kandrosov) [inline AnaProd/tasks.py:121]: can be removed, right?

File: Analysis/SkimEvents.py [hunk 3]
```diff
@@ -19,9 +19,10 @@ def ListToVector(list, type="string"):
   'Long_t' :'long',
   'UInt_t' :'unsigned int',
   'ROOT::VecOps::RVec<float>':'ROOT::VecOps::RVec<float>',
-  'ROOT::VecOps::RVec<int>':'ROOT::VecOps::RVec<int>'
+  'ROOT::VecOps::RVec<int>':'ROOT::VecOps::RVec<int>',
+  'ROOT::VecOps::RVec<unsigned char>':'ROOT::VecOps::RVec<unsigned char>'
   }
-
+#ciao
```
Comment (kandrosov) [inline Analysis/SkimEvents.py:25]:
```suggestion
```

File: Analysis/SkimEvents.py [hunk 4]
```diff
@@ -44,6 +45,8 @@ def make_df(inputFileCentral,inputFileShifted,outDir,treeName):
 
   for var_idx,var_name in enumerate(colNames):
     if var_name in colToNotToMakeDiff: continue
+    #print(var_name)
+    #print(col_types[var_idx])
```
Comment (kandrosov) [inline Analysis/SkimEvents.py:49]:
```suggestion
```

File: Analysis/SkimEvents.py [hunk 5]
```diff
@@ -53,6 +56,7 @@ def make_df(inputFileCentral,inputFileShifted,outDir,treeName):
   df_out_valid_diff = df_out_valid.Filter("!isSame")
   for var_idx,var_name in enumerate(colNames):
     if var_name in colToNotToMakeDiff: continue
+    #print(var_name)
```
Comment (kandrosov) [inline Analysis/SkimEvents.py:59]:
```suggestion
```

File: Analysis/SkimProducer.py [hunk 6]
```diff
@@ -33,21 +36,36 @@
...
+
+    if args.test and k>=1 :
+      #print(systFile)
+      continue
     inFileShiftedName = os.path.join(args.inputDir, systFile)
-    if args.test: print(inFileShiftedName)
+    if args.test: print('shifted file = ', inFileShiftedName)
+    if args.test: print('index = ', k)
     treeName = systFile.strip('.root')
     cmd = f"""python3 /afs/cern.ch/work/v/vdamante/hhbbTauTauRes/prod/Framework/Analysis/SkimEvents.py --inFileCentral {inFileCentralName} --inFileShifted {inFileShiftedName} --outDir {args.workingDir} --treeName {treeName}"""
```
Comment (kandrosov) [inline Analysis/SkimProducer.py:47]: Avoid using absolute paths!

File: Analysis/SkimProducer.py [hunk 7]
```diff
@@ -33,21 +36,36 @@
...
+  hadd_str = f'hadd -f209 -j -O {outFileName} '
   hadd_str += ' '.join(f for f in syst_files_to_merge)
   if args.test: print(hadd_str)
   sh_call([hadd_str], True)
-  shutil.rmtree(args.workingDir)
\ No newline at end of file
+  for file_syst in syst_files_to_merge:
+    if args.test: break
+    os.remove(file_syst)
+  #shutil.rmtree(args.workingDir)
```
Comment (kandrosov) [inline Analysis/SkimProducer.py:71]:
```suggestion
```

File: Common/ConvertUproot.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import awkward as ak
+import os
+import numpy as np
+def saveFile(outFile, out_tree, treeName):
+  if os.path.exists(outFile):
+    os.remove(outFile)
+    with uproot.recreate(outFile, compression=uproot.LZMA(9)) as out_file:
+      out_file[treeName] = out_tree
+      out_file.close()
+    return out_file,outFile
```
Comment (kandrosov) [inline Common/ConvertUproot.py:11]: `out_file` is closed at this point. you shouldn't return it.
```suggestion
return outFile
```

File: Common/ConvertUproot.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+def toUproot(workingDir, fileName):
+  inFile = os.path.join(workingDir, fileName)
+  treeName = fileName.strip(".root")
+  input_file = uproot.open(f"{inFile}")
+  input_tree = input_file[f"{treeName}"]
+  keys = input_tree.keys()
+  outFile=f"{os.path.join(workingDir, treeName)}.root"
+  out_tree = {}
+  if not keys:
+    out_tree = {"noneCol":np.array([-1,-1,-1,-1])}
```
Comment (kandrosov) [inline Common/ConvertUproot.py:22]: Why? If the tree is empty just create an empty output file

File: Common/ConvertUproot.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+  for col_name, columns in collections.items():
+    out_tree[col_name] = ak.zip({ column: df[col_name + "_" + column] for column in columns })
+
+  for column in other_columns:
+    out_tree[column] = df[column]
+
+  #print(out_tree['weight'])
+
+  #print(outFile)
```
Comment (kandrosov) [inline Common/ConvertUproot.py:45]:
```suggestion
```

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: AnaProd/anaTupleProducer.py [hunk 11]
```diff
@@ -64,23 +64,25 @@ def addAllVariables(dfw, syst_name, isData, trigger_class, mode, nLegs):
...
-    dfw.DefineAndAppend(f"ExtraJet_pt", f"v_ops::pt(Jet_p4[ExtraJet_B1])")
-    dfw.DefineAndAppend(f"ExtraJet_eta", f"v_ops::eta(Jet_p4[ExtraJet_B1])")
-    dfw.DefineAndAppend(f"ExtraJet_phi", f"v_ops::phi(Jet_p4[ExtraJet_B1])")
-    dfw.DefineAndAppend(f"ExtraJet_mass", f"v_ops::mass(Jet_p4[ExtraJet_B1])")
-    dfw.DefineAndAppend(f"ExtraJet_ptRes", f"Jet_ptRes[ExtraJet_B1]")
+    #dfw.DefineAndAppend(f"ExtraJet_pt", f"v_ops::pt(Jet_p4[ExtraJet_B1])")
+    #dfw.DefineAndAppend(f"ExtraJet_eta", f"v_ops::eta(Jet_p4[ExtraJet_B1])")
+    #dfw.DefineAndAppend(f"ExtraJet_phi", f"v_ops::phi(Jet_p4[ExtraJet_B1])")
+    #dfw.DefineAndAppend(f"ExtraJet_mass", f"v_ops::mass(Jet_p4[ExtraJet_B1])")
+    #dfw.DefineAndAppend(f"ExtraJet_ptRes", f"Jet_ptRes[ExtraJet_B1]")
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply AnaProd/anaTupleProducer.py:71]: ok

File: AnaProd/anaTupleProducer.py [hunk 12]
```diff
@@ -64,23 +64,25 @@ def addAllVariables(dfw, syst_name, isData, trigger_class, mode, nLegs):
...
     jet_obs = []
     jet_obs.extend(JetObservables)
     dfw.Apply(Baseline.ApplyJetSelection)
     if not isData:
         dfw.Define(f"Jet_genJet_idx", f" FindMatching(Jet_p4,GenJet_p4,0.3)")
         jet_obs.extend(JetObservablesMC)
-        if "LHE_HT" in dfw.df.GetColumnNames():
-            dfw.colToSave.append("LHE_HT")
+        #if "LHE_HT" in dfw.df.GetColumnNames():
+        #    dfw.colToSave.append("LHE_HT")
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply AnaProd/anaTupleProducer.py:79]: ok

File: AnaProd/anaTupleProducer.py [hunk 13]
```diff
@@ -343,7 +353,7 @@ def createAnatuple(inFile, outDir, config, sample_name, anaCache, snapshotOption
     snapshotOptions.fOverwriteIfExists=False
     snapshotOptions.fLazy = True
     snapshotOptions.fMode="RECREATE"
-    snapshotOptions.fCompressionAlgorithm = getattr(ROOT.ROOT, 'k' + args.compressionAlgo)
+    #snapshotOptions.fCompressionAlgorithm = getattr(ROOT.ROOT, 'k' + args.compressionAlgo)
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply AnaProd/anaTupleProducer.py:356]: idr

File: Analysis/SkimEvents.py (hunk 4 above)
Comment (valeriadamante) [reply Analysis/SkimEvents.py:49]: ok

File: Analysis/SkimEvents.py (hunk 5 above)
Comment (valeriadamante) [reply Analysis/SkimEvents.py:59]: ok

File: AnaProd/anaTupleProducer.py [hunk 14]
```diff
@@ -203,10 +206,12 @@ def LegVar(var_name, var_expr, var_type=None, var_cond=None, check_leg_type=True
 
 def createAnatuple(inFile, outDir, config, sample_name, anaCache, snapshotOptions,range, evtIds,
                    store_noncentral, compute_unc_variations, uncertainties, print_cutflow, mode):
+    #print(f"infile = {inFile}, outdir = {outDir}, anaCache = {anaCache}")
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply AnaProd/anaTupleProducer.py:209]: ok

File: AnaProd/tasks.py (hunk 2 above)
Comment (valeriadamante) [reply AnaProd/tasks.py:121]: yes

File: Analysis/SkimProducer.py (hunk 7 above)
Comment (valeriadamante) [reply Analysis/SkimProducer.py:71]: ok

File: Common/ConvertUproot.py (hunk 9 above)
Comment (valeriadamante) [reply Common/ConvertUproot.py:22]: ok

File: Common/ConvertUproot.py (hunk 8 above)
Comment (valeriadamante) [reply Common/ConvertUproot.py:11]: ok

File: Analysis/SkimProducer.py (hunk 6 above)
Comment (valeriadamante) [reply Analysis/SkimProducer.py:47]: ok

File: AnaProd/anaTupleProducer.py [hunk 15]
```diff
@@ -273,7 +278,8 @@ def createAnatuple(inFile, outDir, config, sample_name, anaCache, snapshotOption
             for puIDbranch in puIDbranches:
                 if puIDbranch in dfw.df.GetColumnNames():
                     new_branch_name= puIDbranch.strip("_tmp")
-                    dfw.DefineAndAppend(new_branch_name, f"{puIDbranch}[ExtraJet_B1]")
+                    #dfw.DefineAndAppend(new_branch_name, f"{puIDbranch}[ExtraJet_B1]")
```
Comment (kandrosov) [inline AnaProd/anaTupleProducer.py:281]: don't comment. should be inside `if config["GLOBAL"]["storeExtraJets"]`

File: AnaProd/tasks.py [hunk 16]
```diff
@@ -40,9 +41,15 @@ def run(self):
         print(f'anaCache for sample {sample_name} is created in {self.output().path}')
 
 class AnaTuplePreTask(Task, HTCondorWorkflow, law.LocalWorkflow):
```
Comment (kandrosov) [inline AnaProd/tasks.py:43]: AnaTuplePreTask -> AnaTupleTask

File: AnaProd/tasks.py [hunk 17]
```diff
@@ -94,8 +97,8 @@ def getInputFiles(central_nanoAOD_path, sample_name):
         return list(sorted(input_files))
 
     @staticmethod
-    def getOutputFile(central_anaTuples_path, sample_name, split_idx):
-        return os.path.join(central_anaTuples_path, '_pre', sample_name, f'anaTuple_{split_idx}.root')
+    def getOutputDir(central_anaTuples_path, sample_name):
+        return os.path.join(central_anaTuples_path, '_pre', sample_name)
```
Comment (kandrosov) [inline AnaProd/tasks.py:101]: Remove `_pre` from the output dir. Now these are final anaTuples

File: Analysis/SkimProducer.py [hunk 18]
```diff
@@ -33,21 +36,31 @@
...
-    if args.test: print(inFileCentralName)
+    if args.test and k>=1 :
+      continue
     inFileShiftedName = os.path.join(args.inputDir, systFile)
-    if args.test: print(inFileShiftedName)
+    if args.test: print('shifted file = ', inFileShiftedName)
+    if args.test: print('index = ', k)
     treeName = systFile.strip('.root')
-    cmd = f"""python3 /afs/cern.ch/work/v/vdamante/hhbbTauTauRes/prod/Framework/Analysis/SkimEvents.py --inFileCentral {inFileCentralName} --inFileShifted {inFileShiftedName} --outDir {args.workingDir} --treeName {treeName}"""
+    cmd = f"""python3 Analysis/SkimEvents.py --inFileCentral {inFileCentralName} --inFileShifted {inFileShiftedName} --outDir {args.workingDir} --treeName {treeName}"""
```
Comment (kandrosov) [inline Analysis/SkimProducer.py:45]: put full path for py starting with ANALYSIS_PATH

File: Common/ConvertUproot.py [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+  input_tree = input_file[f"{treeName}"]
+  keys = input_tree.keys()
+  outFile=f"{os.path.join(workingDir, treeName)}.root"
+  out_tree = {}
+  if not keys:
+    with uproot.recreate(outFile, compression=uproot.LZMA(9)) as out_file:
+      out_file.close()
+    return outFile
+    #out_tree = {"noneCol":np.array([-1,-1,-1,-1])}
+    #return saveFile(outFile, out_tree, treeName)
```
Comment (kandrosov) [inline Common/ConvertUproot.py:26]:
```suggestion
```

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: AnaProd/tasks.py (hunk 16 above)
Comment (valeriadamante) [reply AnaProd/tasks.py:43]: ok

File: AnaProd/tasks.py (hunk 17 above)
Comment (valeriadamante) [reply AnaProd/tasks.py:101]: ok

File: Analysis/SkimProducer.py (hunk 18 above)
Comment (valeriadamante) [reply Analysis/SkimProducer.py:45]: ok

File: Common/ConvertUproot.py (hunk 19 above)
Comment (valeriadamante) [reply Common/ConvertUproot.py:26]: ok

File: AnaProd/anaTupleProducer.py (hunk 15 above)
Comment (valeriadamante) [reply AnaProd/anaTupleProducer.py:281]: ok

File: AnaProd/anaTupleProducer.py [hunk 20]
```diff
@@ -273,7 +277,7 @@ def createAnatuple(inFile, outDir, config, sample_name, anaCache, snapshotOption
             for puIDbranch in puIDbranches:
                 if puIDbranch in dfw.df.GetColumnNames():
                     new_branch_name= puIDbranch.strip("_tmp")
-                    dfw.DefineAndAppend(new_branch_name, f"{puIDbranch}[ExtraJet_B1]")
+                    dfw.Define(new_branch_name, f"{puIDbranch}[ExtraJet_B1]")
```
Comment (kandrosov) [inline AnaProd/anaTupleProducer.py:280]: In this way it will start with `weight_`, isn't it? we want it to start with `ExtraJet_weight_`.

**Review by kandrosov — CHANGES_REQUESTED** [review]
update Corrections

**Review by kandrosov — APPROVED** [review]

**Pull request merged**